README.md
<p align="center">
  <img src="https://phage.ai/assets/images/phageai_logo2.svg">
</p>

**PhageAI** is an application that simultaneously represents **a repository of knowledge of bacteriophages** and a bioinformatics pipeline to analyse genomes with **Artificial Intelligence support**. This package supports the most critical programmable features from our platform.

Machine Learning algorithms can process enormous amounts of data in relatively short time in order to find connections and dependencies that are obvious for human beings. Correctly designed applications based on AI are able to vastly improve and speed up the work of the domain experts.

Models based on DNA or proteins contextual vectorization and Deep Neural Networks are particularly effective when it comes to analysis of genomic data. The system that we propose aims to use the phages sequences uploaded to the database to build a model which is able to predict if a bacteriophage is **chronic**, **temperate** or **virulent** with a high probability. Furthermore, our system shares more prediction methods for phage taxonomy, phage similarity and annotation extended by proteins structural classes classification.

One of the key system modules is the bacteriophages repository with a clean web interface that allows to browse, upload and share data with other users. The gathered knowledge about the bacteriophages is not only valuable on its own but also because of the ability to train the ever-improving Machine Learning models.

Detection of virulent or temperate features is only one of the first tasks that can be solved with Artificial Intelligence. The combination of Biology, Natural Language Processing and Machine Learning allows us to create algorithms for genomic data processing that could eventually turn out to be effective in a wide range of problems with focus on classification and information extraced from DNA.


[![PyPI version](https://img.shields.io/pypi/v/phageai.svg)](https://pypi.org/project/phageai/)
[![PyPI license](https://img.shields.io/pypi/l/phageai.svg)](https://pypi.python.org/pypi/)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/phageai.svg)](https://pypi.python.org/pypi/phageai/)
[![Code style](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![Downloads](https://static.pepy.tech/badge/phageai)](https://pepy.tech/project/phageai)
[![Twitter Follow](https://img.shields.io/twitter/follow/phageai.svg?style=social)](https://twitter.com/phageai)
[![LinkedIn Follow](https://img.shields.io/badge/LinkedIn-0077B5?style=flat&logo=appveyor)](https://pl.linkedin.com/company/phageai-s-a)


## Table of Contents

[Available methods](https://github.com/phageaisa/phageai#available-methods) | [Installation](https://github.com/phageaisa/phageai#installation-and-usage) | [Benchmark](https://github.com/phageaisa/phageai#benchmark) | [Community and Contributions](https://github.com/phageaisa/phageai#community-and-contributions) | [Have a question?](https://github.com/phageaisa/phageai#have-a-question) | [Found a bug?](https://github.com/phageaisa/phageai#found-a-bug) | [Team](https://github.com/phageaisa/phageai#team) | [Change log](https://github.com/phageaisa/phageai#change-log) | [License](https://github.com/phageaisa/phageai#license) | [Cite](https://github.com/phageaisa/phageai#cite)

## Available methods

* `upload(fasta_path, access)` - upload FASTA file with phage genome as "public", "private" or "" (temporary) sample in the PhageAI repository. Upload stage is starting the bioinformatics pipeline execution for phage characteristic.
* `processing_status(job_id)` - get current processing status for your phage sample related with Job ID;
* `get_lifecycle_classification(job_id)` - get phage lifecycle classification result;
* `get_taxonomy_classification(job_id)` - get phage taxonomy classification results for order, family and genus;
* `get_proteins_classification(job_id)` - get phage proteins structural classes classification results;
* `get_top10_similarities(job_id)` - get TOP-10 similar phages to your sample from the repository;
* `get_full_report(job_id)` - get full phage characteristics report (all meta-data and predictions);
* `get_phage_characteristic(accession_number)` - get meta-data about publicly available phage with specific accession number (with version);


## Installation and usage

#### PhageAI user account (1/3)
Create a free user account in [the PhageAI web platform](https://app.phage.ai/) or use an existing one. If you had to create new one, activate your account by activation link which was sent on your mail inbox. After that, log into the platform successfully and click "My profile" in the top-right menu. In the "API access" section create a new access token (string) and copy it for the steps below.

<p align="center">
  <img src="https://raw.githubusercontent.com/phageaisa/phageai/main/media/phageai-access-token.png">
</p>

#### PhageAI package (2/3)

_PhageAI_ requires Python 3.8.0+ to run and can be installed by running:

```
pip install phageai
```

If you can't wait for the latest hotness from the develop branch, then install it directly from the repository:

```
pip install git+https://github.com/phageaisa/phageai.git@develop
```

#### PhageAI execution (3/3)

`PASTE_YOUR_ACCESS_TOKEN_HERE` - PhageAI web user's access token;
`PASTE_YOUR_FASTA_PATH_HERE` - FASTA filename with *.fasta or *.fa extension;

### Example - how to upload phage to repository with private access

```python
from phageai.platform import PhageAIAccounts

phageai_api = PhageAIAccounts(access_token='PASTE_YOUR_ACCESS_TOKEN_HERE')

phage_example_jobid = phageai_api.upload(fasta_path="PASTE_YOUR_FASTA_PATH_HERE", access="private")
```

Expected output should be the job ID value:
```json
{
  'job_id': '0a71e61a-ec58-447b-859e-d9ba15e103a9'
}
```

or, if you reach out daily API requests limit (100 by default), you can expect:

```json
{
    "author": ["Your daily API limit (100 requests) has been exceeded"]
}
```

If you reach out your daily requests limit, and you still need more, feel free to contact us by contact@phageai.

### Example - how to track the processing progress

Tracking progress of phage processing is super useful if you upload more samples in the same time or when you integrate your service or pipeline with PhageAI.

```python
(...)

phage_example_jobid = phageai_api.upload(fasta_path="PASTE_YOUR_FASTA_PATH_HERE", access="private")

job_id = phage_example_jobid["job_id"]

phageai_api.processing_status(job_id=job_id)
```

Depends on what is the current stage of your sample processing ("Not Started", "In progress", "Done", "Failed"), expected output should be like:

```python
{
  'taxonomy_stage': 'Done',
  'proteins_stage': 'Not Started',
  'top10_stage': 'Not Started',
  'lifecycle_stage': 'Done',
  'final_report': 'In Progress'
}
```

Each of the above stage is work separately in PhageAI so you can expect different statuses for each of them.

### Example - get lifecycle classification

```python
(...)

phageai_api.get_lifecycle_classification(job_id=job_id)
```

Expected output should be like:

```python
{'value': 'Chronic', 'probability': 99.85}
```

In the same way you can execute other [available methods](https://github.com/phageaisa/phageai#available-methods) from the package. 

## Benchmark

PhageAI lifecycle classifier was benchmarked with [DeePhage](https://github.com/shufangwu/DeePhage), [bacphlip](https://github.com/adamhockenberry/bacphlip), [VIBRANT](https://github.com/AnantharamanLab/VIBRANT) and [PHACTS](https://github.com/deprekate/PHACTS) tools using 104 Virulent and Temperate bacteriophages from our paper (testing set). Correct predictions results:

Tool | Version   | Chronic support | Phage sequences used in research | Test set accuracy (%) | DOI |
--- |-----------| --- |----------------------------------|-----------------------| --- |
**PhageAI** | **0.4.1** | **Yes** | **15,235**                       | **93.27**             | **This research** |
DeePhage | 1.0       | No | 1,640                            | 84.62                 | 10.1093/gigascience/giab056 |
bacphlip | 0.9.6     | No | 1,057                            | 100                   | 10.7717/peerj.11396 |
VIBRANT | 1.2.1     | No | 350,626                          | 85.58                 | 10.1186/s40168-020-00867-0 |
PHACTS | 1.8       | No | 227                              | 75.00                 | 10.1093/bioinformatics/bts014 |

## Community and Contributions

Happy to see you willing to make the PhageAI better. Development on the latest stable version of Python 3+ is preferred. As of this writing it's 3.8. You can use any operating system.

If you're fixing a bug or adding a new feature, add a test with *[pytest](https://github.com/pytest-dev/pytest)* and check the code with *[Black](https://github.com/psf/black/)* and *[mypy](https://github.com/python/mypy)*. Before adding any large feature, first open an issue for us to discuss the idea with the core devs and community.

## Have a question?

Obviously if you have a private question or want to cooperate with us, you can always reach out to us directly by mail.

## Found a bug?

Feel free to add a new issue with a respective title and description on the [the PhageAI repository](https://github.com/phageaisa/phageai/issues). If you already found a solution to your problem, we would be happy to review your pull request.

## Team

Core Developers and Domain Experts who contributing to PhageAI:

* Piotr Tynecki
* Łukasz Wałejko
* Krzysztof Owsieniuk
* Joanna Kazimierczak
* Arkadiusz Guziński
* Bogumił Zimoń
* Żaneta Szulc
* Maria Urbanowicz

## Change log

The log's will become rather long. It moved to its own file.

See [CHANGELOG.md](https://github.com/phageaisa/phageai/blob/master/CHANGELOG.md).

## License

The PhageAI package is released under the under terms of [the MIT License](https://github.com/phageaisa/phageai/blob/master/LICENSE).

## Cite

> **PhageAI - Bacteriophage Life Cycle Recognition with Machine Learning and Natural Language Processing**
>
> Tynecki, P.; Guziński, A.; Kazimierczak J.; Zimoń B.; Szulc Ż.; Jadczuk, M.; Dastych, J.; Onisko, A.
>
> Viruses, Special Issue "Bacteriophage Bioinformatics"
(ISSN 1999-4915), DOI: [10.1101/2020.07.11.198606](https://doi.org/10.1101/2020.07.11.198606)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.0.0] - 15.11.2023
### Changed
* Reimplemented connector to the PhageAI API; 
* Added support to taxonomy classification (order, family, genus), top-10 similarities, annotation and proteins class prediction, full phage characteristic report and phage repository;
* Removed support for `LifeCycleClassifier` and `BacteriophageRepository` classes;
* `README.md` updated; 


## [0.4.2] - 29.09.2023
### Changed
* `README.md` updated;


## [0.4.1] - 29.09.2022
### Changed
* `datasets` files updated;
* `README.md` updated;


## [0.4.0] - 02.08.2022
### Changed
* Refactored PhageAI API connectors;


## [0.3.1] - 24.02.2022
### Changed
* Bugfix for URL addresses;


## [0.3.0] - 22.10.2021
### Added
* Added method for fetching bacteriophage meta-data;
* Added method for calculating alignment-free phage similarity (top-10);
### Changed
* `README.md` updated;


## [0.2.0] - 03.08.2021
### Added
* Added user requests limit support;
* Added chronic lifecycle support;
### Changed
* `README.md` updated;


## [0.1.1] - 03.12.2020
### Changed
* `README.md` updated;


## [0.1.0] - 12.11.2020
### Added
* Added `LifeCycleClassifier` support to istinguish between the two major types of phages life cycles: lytic and lysogenic;


phageai/platform/tests/phageai_api_tests.py
import uuid
from unittest import mock

import pytest

from phageai.platform import PhageAIAccounts


def test_invalid_access_token_and_fasta_file():
    """
    Test API reaction on invalid access token and FASTA file
    """

    phageai_api = PhageAIAccounts(access_token="INVALID_TOKEN")

    result = phageai_api.upload(fasta_path="INVALID_FASTA_FILE")

    assert result == {}


def test_old_access_token():
    """
    Test API reaction on invalid access token and FASTA file
    """

    with pytest.raises(ValueError) as exc_info:
        PhageAIAccounts(access_token=str(uuid.uuid4()))
    assert "[PhageAI] Token Error: We have change our TOS and Policy." in str(
        exc_info.value
    )


def test_empty_access_token():
    """
    Test API reaction on empty access token
    """

    with pytest.raises(ValueError) as exc_info:
        PhageAIAccounts(access_token="")
    assert "[PhageAI] Token Error: Please provide correct access token." in str(
        exc_info.value
    )


def test_null_access_token():
    """
    Test API reaction on null access token
    """

    with pytest.raises(ValueError) as exc_info:
        PhageAIAccounts(access_token=None)
    assert "[PhageAI] Token Error: Please provide correct access token." in str(
        exc_info.value
    )


def test_correct_fasta_file():
    """
    Test API reaction on valid FASTA file
    """

    response_example = {
        "lifecycle": {"value": "Virulent", 'probability': 100.0},
        "gc": "41.45",
        "length": 171598
    }

    with mock.patch(
        "phageai.platform.PhageAIAccounts._make_request"
    ) as mock_connector:
        mock_connector.return_value.json.return_value = response_example

        phageai_api = PhageAIAccounts(access_token="GOOD_TOKEN")
        result = phageai_api.upload(fasta_path="phageai/platform/tests/NC_055712.fasta")

        assert result == response_example


def test_incorrect_fasta_file():
    """
    Test API reaction on invalid FASTA file
    """

    response_example = {}

    with mock.patch(
        "phageai.platform.PhageAIAccounts._make_request"
    ) as mock_connector:
        mock_connector.return_value.json.return_value = response_example

        phageai_api = PhageAIAccounts(access_token="GOOD_TOKEN")
        result = phageai_api.upload(fasta_path="BAD_FILE.fasta")

        assert result == response_example


phageai/lifecycle/tests/test_lifecycle_executions.py
import json
import uuid
from unittest import mock

import pytest

from phageai.lifecycle.classifier import LifeCycleClassifier


def test_invalid_access_token_and_fasta_file():
    """
    Test API reaction on invalid access token and FASTA file
    """

    lcc = LifeCycleClassifier(access_token="INVALID_TOKEN")

    result = lcc.predict(fasta_path="INVALID_FASTA_FILE")

    assert result == {}


def test_old_access_token():
    """
    Test API reaction on invalid access token and FASTA file
    """

    with pytest.raises(ValueError) as exc_info:
        LifeCycleClassifier(access_token=str(uuid.uuid4()))
    assert "[PhageAI] Token Error: We have change our TOS and Policy." in str(
        exc_info.value
    )


def test_empty_access_token():
    """
    Test API reaction on empty access token
    """

    with pytest.raises(ValueError) as exc_info:
        LifeCycleClassifier(access_token="")
    assert "[PhageAI] Token Error: Please provide correct access token." in str(
        exc_info.value
    )


def test_null_access_token():
    """
    Test API reaction on null access token
    """

    with pytest.raises(ValueError) as exc_info:
        LifeCycleClassifier(access_token=None)
    assert "[PhageAI] Token Error: Please provide correct access token." in str(
        exc_info.value
    )


def test_correct_fasta_file():
    """
    Test API reaction on valid FASTA file
    """

    response_example = {
        "hash": "4b0b12e3e4af0d79ba172240fa20bf2f912e2297e3ab1c3b7207200877e7ceab",
        "predicted_lifecycle": "Virulent",
        "gc": "41.45",
        "prediction_accuracy": "100.00",
        "sequence_length": 171599,
    }

    with mock.patch(
        "phageai.lifecycle.classifier.LifeCycleClassifier._make_request"
    ) as mock_connector:
        mock_connector.return_value.json.return_value = response_example

        lcc = LifeCycleClassifier(access_token="GOOD_TOKEN")
        result = lcc.predict(fasta_path="phageai/lifecycle/tests/NC_055712.fasta")

        assert result == response_example


def test_incorrect_fasta_file():
    """
    Test API reaction on invalid FASTA file
    """

    response_example = {}

    with mock.patch(
        "phageai.lifecycle.classifier.LifeCycleClassifier._make_request"
    ) as mock_connector:
        mock_connector.return_value.json.return_value = response_example

        lcc = LifeCycleClassifier(access_token="GOOD_TOKEN")
        result = lcc.predict(fasta_path="BAD_FILE.fasta")

        assert result == response_example


phageai/repository/tests/test_repository_executions.py
import uuid
from unittest import mock

import pytest

from phageai.repository.phages import BacteriophageRepository


def test_invalid_access_token_and_fasta_file_1():
    """
    Test API reaction on invalid access token and FASTA file
    """

    lcc = BacteriophageRepository(access_token="INVALID_TOKEN")

    result = lcc.get_record(value="NC_000866")

    assert result == []


def test_invalid_access_token_and_fasta_file_2():
    """
    Test API reaction on invalid access token and FASTA file
    """

    lcc = BacteriophageRepository(access_token="INVALID_TOKEN")

    result = lcc.get_top10_similar_phages(value="NC_000866")

    assert result == []


def test_old_access_token():
    """
    Test API reaction on invalid access token and FASTA file
    """

    with pytest.raises(ValueError) as exc_info:
        BacteriophageRepository(access_token=str(uuid.uuid4()))
    assert "[PhageAI] Token Error: We have change our TOS and Policy." in str(
        exc_info.value
    )


def test_empty_access_token():
    """
    Test API reaction on empty access token
    """

    with pytest.raises(ValueError) as exc_info:
        BacteriophageRepository(access_token="")
    assert "[PhageAI] Token Error: Please provide correct access token." in str(
        exc_info.value
    )


def test_null_access_token():
    """
    Test API reaction on null access token
    """

    with pytest.raises(ValueError) as exc_info:
        BacteriophageRepository(access_token=None)
    assert "[PhageAI] Token Error: Please provide correct access token." in str(
        exc_info.value
    )


def test_correct_accession_number():
    """
    Test API reaction on valid accession number
    """

    response_example = [
        {
            "hash": "edab9a1ed79c0436490cb569449b8ecbd03e591f45510c169df5076226e81b66",
            "accession_no": "MZ375324",
            "name": "Salmonella phage seszw",
            "description": "Mutant salmonella phage seszw clone seszw_150s, complete genome",
            "host": "Salmonella",
            "host_full": "Salmonella",
            "origin_lifecycle": "Unknown",
            "origin_order": "",
            "origin_family": "",
            "origin_genus": "",
            "predicted_lifecycle": {"name": "Temperate", "accuracy": "93.50"},
            "predicted_order": {"name": "Caudovirales", "accuracy": "99.85"},
            "predicted_family": {"name": "Siphoviridae", "accuracy": "98.59"},
            "predicted_genus": {"name": "Roufvirus", "accuracy": "99.20"},
            "sequence_length": 40462,
            "gc": "46.16",
            "genome_coverage_label": "Complete genome",
            "genome_composition_label": "Unknown",
            "is_refseq": "No",
        }
    ]

    with mock.patch(
        "phageai.repository.phages.BacteriophageRepository._make_request"
    ) as mock_connector:
        mock_connector.return_value.json.return_value = response_example

        lcc = BacteriophageRepository(access_token="GOOD_TOKEN")
        result = lcc.get_record(value="MZ375324")

        assert result == response_example


def test_incorrect_accession_number():
    """
    Test API reaction on invalid accession number
    """

    response_example = []

    with mock.patch(
        "phageai.repository.phages.BacteriophageRepository._make_request"
    ) as mock_connector:
        mock_connector.return_value.json.return_value = response_example

        lcc = BacteriophageRepository(access_token="GOOD_TOKEN")
        result = lcc.get_record(value="BAD_ACCNO")

        assert result == response_example


phageai/phageai_auth.py
import logging
import uuid

import requests

from urllib.parse import urljoin

from . import PHAGEAI_BASE_URL

logging.basicConfig(level=logging.INFO)


class PhageAIConnector:
    """
    Generic PhageAI API connector
    """

    REQUEST_TIMEOUT = 30

    API_URL = urljoin(PHAGEAI_BASE_URL, "/api/v1/phageai-package/")

    def __init__(self, access_token: str) -> None:
        if not access_token:
            raise ValueError(
                "[PhageAI] Token Error: Please provide correct access token. If you need more information, please check README."
            )
        if self._is_uuid(access_token):
            raise ValueError(
                "[PhageAI] Token Error: We have change our TOS and Policy. Please login to PhageAI Web platform (https://app.phage.ai/) and create new access token."
            )
        self.access_token = access_token
        self.result = {}

    @staticmethod
    def _is_uuid(value):
        try:
            uuid.UUID(value)
            return True
        except ValueError:
            return False

    def _create_auth_header(self) -> dict:
        return {
            "Authorization": f"Bearer {self.access_token}"
        }

    @staticmethod
    def _check_status(response):
        http_error_msg = ""
        if isinstance(response.reason, bytes):
            # We attempt to decode utf-8 first because some servers
            # choose to localize their reason strings. If the string
            # isn't utf-8, we fall back to iso-8859-1 for all other
            # encodings. (See PR #3538)
            try:
                reason = response.reason.decode("utf-8")
            except UnicodeDecodeError:
                reason = response.reason.decode("iso-8859-1")
        else:
            reason = response.reason

        if 400 <= response.status_code < 500:
            http_error_msg = f"[PhageAI] Connection Client Error: Response Status Code - {response.status_code} Reason: {reason}"
        elif 500 <= response.status_code < 600:
            http_error_msg = f"[PhageAI] Connection Server Error: Response Status Code - {response.status_code} Reason: {reason}"

        if http_error_msg:
            raise requests.HTTPError(http_error_msg, response=response)

    def _make_request(self, path: str, method: str, **kwargs) -> requests.Response:
        """
        Generic PhageAI API request method
        """

        headers = self._create_auth_header()

        response = getattr(requests, method)(
            url=urljoin(self.API_URL, path),
            headers=headers,
            timeout=self.REQUEST_TIMEOUT,
            **kwargs
        )

        self._check_status(response)

        return response


phageai/repository/phages.py
import logging
from typing import List

import requests

from phageai.phageai_auth import PhageAIConnector


class BacteriophageRepository(PhageAIConnector):
    EXPECTED_HTTP_STATUS = 200
    PATH = "YmFjdGVyaW9waGFnZS8="

    def get_record(self, value: str) -> List[dict]:
        """
        Return dict with bacteriophage meta-data
        """

        result = []

        try:
            response = self._make_request(
                path=self._encode(f"{self._decode(self.PATH)}{value}/"), method="get"
            )

            result = response.json()

            if response.status_code == self.EXPECTED_HTTP_STATUS:
                logging.info(f"[PhageAI] Phage get record executed successfully")
            else:
                logging.warning(f'[PhageAI] Exception was raised: "{result}"')
        except requests.exceptions.RequestException as e:
            logging.warning(f'[PhageAI] Exception was raised: "{e}"')

        return result

    def get_top10_similar_phages(self, value: str) -> List[dict]:
        """
        Return list of dicts contained top-10 most similar bacteriophages
        """

        result = []

        try:
            response = self._make_request(
                path=self._encode(
                    f"{self._decode(self.PATH)}{value}{self._decode('L3RvcC0xMC8=')}"
                ),
                method="get",
            )

            result = response.json()

            if response.status_code == self.EXPECTED_HTTP_STATUS:
                logging.info(
                    f"[PhageAI] Phage top-10 similar phages executed successfully"
                )
            else:
                logging.warning(f'[PhageAI] Exception was raised: "{result}"')
        except requests.exceptions.RequestException as e:
            logging.warning(f'[PhageAI] Exception was raised: "{e}"')

        return result


phageai/lifecycle/classifier.py
import os

import requests
import logging

from phageai.phageai_auth import PhageAIConnector


class LifeCycleClassifier(PhageAIConnector):
    """
    Bacteriophage life cycle classifier

    All the research and scientific details were published in the paper:
    DOI: 10.1101/2020.07.11.198606
    """

    PATH = "bGlmZWN5Y2xlX3ByZWRpY3Rpb24v"
    EXPECTED_HTTP_STATUS = 201

    def predict(self, fasta_path: str) -> dict:
        """
        Return dict structure with predicted class (label), prediction accuracy, GC% and sequence length
        for passed bacteriophage FASTA file
        """

        result = {}

        if os.path.exists(fasta_path):
            with open(fasta_path, "rb") as fasta:
                try:
                    response = self._make_request(
                        path=self.PATH,
                        method="post",
                        files=[("file", fasta)],
                    )

                    result = response.json()

                    if response.status_code == self.EXPECTED_HTTP_STATUS:
                        logging.info(
                            f"[PhageAI] Life cycle classifier executed successfully"
                        )
                    else:
                        logging.warning(f'[PhageAI] Exception was raised: "{result}"')
                except requests.exceptions.RequestException as e:
                    logging.warning(f'[PhageAI] Exception was raised: "{e}"')
        else:
            logging.warning(
                f'[PhageAI] Exception was raised: "{fasta_path}" doesn\'t exists'
            )

        return result


phageai/platform/phageai_api.py
import os

import requests
import logging
from http import HTTPStatus

from phageai.phageai_auth import PhageAIConnector


class PhageAIAccounts(PhageAIConnector):
    def upload(self, fasta_path: str, access: str = "private") -> dict:
        """
        Upload FASTA file as "public", "private" or "" (temporary) sample in the PhageAI repository
        - "public" means that all the users will be able to see and download the sample;
        - "private" means that only you (as user) will be able to see and download the sample;
        - "" (temporary) means that the sample will be not hosted in the platform longer than during the processing;

        Upload stage is starting the pipeline execution for phage characteristic

        Method returns Job ID value which represents phage processing ID in the platform
        """

        job_id = {}

        if os.path.exists(fasta_path):
            with open(fasta_path, "rb") as fasta:
                try:
                    response = self._make_request(
                        path="upload/",
                        method="post",
                        data={"access": access},
                        files=[("fasta", fasta)]
                    )

                    job_id = response.json()

                    # HTTP 201
                    if response.status_code == HTTPStatus.CREATED:
                        logging.info(
                            f"[PhageAI] Phage upload executed successfully"
                        )
                    else:
                        logging.warning(f'[PhageAI] Exception was raised: "{job_id}"')
                except requests.exceptions.RequestException as e:
                    logging.warning(f'[PhageAI] Exception was raised: "{e}"')
        else:
            logging.warning(
                f'[PhageAI] Exception was raised: "{fasta_path}" doesn\'t exists'
            )

        return job_id

    def processing_status(self, job_id: str):
        """
        Method returns processing status for Phage sample related with Job ID
        """

        processing_results = {}

        try:
            response = self._make_request(
                path=f"upload/{job_id}/status/",
                method="get"
            )

            processing_results = response.json()

            # HTTP 200
            if response.status_code == HTTPStatus.OK:
                logging.info(
                    f"[PhageAI] Phage processing status"
                )
            else:
                logging.warning(f'[PhageAI] Exception was raised: "{processing_results}"')
        except requests.exceptions.RequestException as e:
            logging.warning(f'[PhageAI] Exception was raised: "{e}"')

        return processing_results

    def _execute_phage_method(self, phageai_method: str, value: str, http_method: str):
        """
        Private method for generic phage method execution
        """

        results = {}

        try:
            response = self._make_request(
                path=f"phage/{value}/{phageai_method}/",
                method=http_method
            )

            results = response.json()

            # HTTP 200
            if response.status_code == HTTPStatus.OK:
                logging.info(
                    f"[PhageAI] Phage {phageai_method} method executed successfully"
                )
            else:
                logging.warning(f'[PhageAI] Exception was raised: "{results}"')
        except requests.exceptions.RequestException as e:
            logging.warning(f'[PhageAI] Exception was raised: "{e}"')

        return results

    def get_taxonomy_classification(self, job_id: str):
        """
        Method returns phage taxonomy classification results for order, family and genus
        """

        return self._execute_phage_method(
            phageai_method="taxonomy_classification",
            value=job_id,
            http_method="get",
        )

    def get_proteins_classification(self, job_id: str):
        """
        Method returns phage proteins structural classes classification results
        """

        return self._execute_phage_method(
            phageai_method="proteins_classification",
            value=job_id,
            http_method="get",
        )

    def get_top10_similarities(self, job_id: str):
        """
        Method returns TOP-10 similar phages in the PhageAI repository
        """

        return self._execute_phage_method(
            phageai_method="top10_similarities",
            value=job_id,
            http_method="get",
        )

    def get_lifecycle_classification(self, job_id: str):
        """
        Method returns lifecycle classification result
        """

        return self._execute_phage_method(
            phageai_method="lifecycle_classification",
            value=job_id,
            http_method="get",
        )

    def get_full_report(self, job_id: str):
        """
        Method returns full phage characteristics report
        """

        return self._execute_phage_method(
            phageai_method="report",
            value=job_id,
            http_method="get",
        )

    def get_phage_characteristic(self, accession_number: str):
        """
        Method returns meta-data about phage with specific accession number (with version)
        """

        return self._execute_phage_method(
            phageai_method="characteristic",
            value=accession_number,
            http_method="get",
        )
